src/txmempool.cpp

```diff
@@ -1245,3 +1536,169 @@ std::vector<CTxMemPool::txiter> CTxMemPool::GatherClusters(const std::vector<uin
     }
     return clustered_txs;
 }
+
+void Cluster::AddTransaction(const CTxMemPoolEntry& entry, bool sort)
+{
+    m_chunks.emplace_back(entry.GetModifiedFee(), entry.GetTxSize());
+    m_chunks.back().txs.emplace_back(entry);
+    entry.m_cluster = this;
+    ++m_tx_count;
+    if (sort) Sort();
+    return;
+}
+
+void Cluster::RemoveTransaction(const CTxMemPoolEntry& entry)
+{
+    m_chunks[entry.m_loc.first].txs.erase(entry.m_loc.second);
+
+    // Chunk (or cluster) may now be empty, but this will get cleaned up
+    // when the cluster is re-sorted (or when the cluster is deleted) Note:
+    // if we cleaned up empty chunks here, then this would break the
+    // locations of other entries in the cluster. Since we would like to be
+    // able to do multiple removals in a row and then clean up the sort, we
+    // can't clean up empty chunks here.
+    --m_tx_count;
+    return;
+}
+
+void Cluster::RechunkFromLinearization(std::vector<CTxMemPoolEntry::CTxMemPoolEntryRef>& txs, bool reassign_locations)
+{
+    m_chunks.clear();
+
+    for (auto txentry : txs) {
+        m_chunks.emplace_back(txentry.get().GetModifiedFee(), txentry.get().GetTxSize());
+        m_chunks.back().txs.emplace_back(txentry);
```

Comment:
54f39ca8f101483f5f82707689ca49431d4091e5: So you're creating a chunk for each new transaction and then erasing it if the fee rate goes down. Why not the other way around?